Data Preparation Documentation (#4569)
SoftwareAG/c8y-docs: content/data-preparation/data-preparation-introduction-bundle/overview.md

---
weight: 10
title: Overview
layout: redirect
---

### Data Preparation {#datapreparation}

> The Data Preparation application provides a modern, AI-first environment for creating and managing data transformation logic to help you convert raw device data into the Cumulocity [data model](/concepts/domain-model). As IoT devices often communicate in various formats (from standard JSON to IoT-specific binary protocols), Data Preparation acts as a bridge that ensures your data is standardized, corrected, and ready for use across the platform and downstream.

Reviewer: ```suggestion
The Data Preparation application provides a modern, AI-first environment for creating and managing data transformation logic to help you convert raw device data into the {{< product-c8y-iot >}} [data model](/concepts/domain-model). As IoT devices often communicate in various formats (from standard JSON to IoT-specific binary protocols), Data Preparation acts as a bridge that ensures your data is standardized, corrected, and ready for use across the platform and downstream.
```

Reviewer: Here and in other places in this file.

Reviewer 2: bee43ba76

Data Preparation uses smart functions --- modular pieces of logic that independently process incoming messages to generate one or more Cumulocity-compliant outputs. For details, see [Smart functions](/concepts/smart-function-concept/).

### Why use Data Preparation {#why-use-data-preparation}

Data Preparation empowers you to:

* Easily convert raw payloads into standard Cumulocity measurements, events, alarms, and inventory objects.

> * Use a conversational AI chat interface to describe your business context and automatically generate the necessary transformation code in a Smart Function.

Reviewer: ```suggestion
* Use a conversational AI chat interface to describe your business context and automatically generate the necessary transformation code in a smart function.
```

Reviewer 2: bee43ba76

* Perform real-time calculations (for example, converting Fahrenheit to Celsius) or correct values based on predefined normal ranges.
* Automatically map and [create devices](/data-preparation/device-onboarding/) based on external IDs found in the payload, source client ID, or topic path.
* Scale with support for high-volume data ingestion, as Data Preparation is built on high-performant, scalable infrastructure.

> ### Key Capabilities {#key-capabilities}

Reviewer: ```suggestion
### Key capabilities {#key-capabilities}
```

Reviewer 2: bee43ba76

> * AI-first experience --- the primary user interface is an AI assistant that writes and optimizes Javascript-based transformation logic based on your prompts (leveraging the [AI Agent Manager](/ai/aim-introduction/)).

Reviewer: In other places, we usually start with a capital letter after the dash/hyphen. Again, this is a general formatting question. We can leave this for now as is, as you are very consistent, but I will take this to my list.

Reviewer 2: When just using an emdash in prose (similar to when you would use a comma), then it's definitely lower case---it's just part of the sentence (like this one). Here it's more like a colon: You could include a capital, if it's a complete standalone sentence (like this one). I think in this specific case a capital is correct.

Reviewer 2: 7a5197961

* Built-in code editor --- a simplified IDE is available to manually view, edit, or paste pre-written logic.
* Testing and validation --- run tests using sample data (either manually uploaded or captured live from an MQTT topic) with a visual comparison.
* Integrated deployment --- once a rule is active, it runs continuously as data is posted to the subscribed [MQTT Service](/device-integration/mqtt-service/) topics.  

{{< c8y-admon-important >}}
Data Preparation handles data normalization only. For complex event processing, aggregations, or real-time analytics IoT use cases, use [Streaming Analytics](/streaming-analytics/introduction-analytics/) after Data Preparation has normalized your data.
{{< /c8y-admon-important >}}

### Architecture {#architecture}

The diagram below illustrates the Data Preparation service flows within a tenant.

<p align="center" width="100%">
    <img width="80%" src="/images/data-preparation/datprep_architecture.png" alt="Data Preparation Service architecture">
</p>

>  Data Preparation receives raw device messages, applies user-defined transformation logic, and forwards the resulting Cumulocity objects to the platform for persistence and use by applications (e.g. Streaming Analytics).  

Reviewer: ```suggestion
 Data Preparation receives raw device messages, applies user-defined transformation logic, and forwards the resulting {{< product-c8y-iot >}} objects to the platform for persistence and use by applications (for example, Streaming Analytics).  
```

Reviewer 2: bee43ba76

> ### How Data Preparation works {#how-it-works}

Reviewer: ```suggestion
### How data preparation works {#how-it-works}
```

Reviewer 2: Data Preparation is being used as a proper noun in this title, the same as the previous paragraph. Either "Preparation" should be "preparation" in the previous paragraph, or both of them should be capitalized in the title. I believe the standard for names of applications (like Streaming Analytics) is to always capitalize them (as opposed to features within them, like smart rules, or smart functions, which are lower case)

Reviewer: Yes, my fault, this is used as a proper name throughout the entire text.

Data Preparation listens for incoming device messages on [MQTT Service](/device-integration/mqtt-service/) topics. When a message arrives, it evaluates all active rules subscribed to patterns that match the message topic. Each matching rule runs its smart functions against the payload and the resulting Cumulocity objects --- measurements, events, alarms, or managed objects --- are forwarded to the platform and persisted.

Multiple active rules can subscribe to patterns that match the same topic and execute independently. A single message can trigger multiple rules, and each rule can produce multiple output objects.

### Key concepts {#key-concepts}

> ### Smart Functions {#smart-functions}

Reviewer: ```suggestion
#### Smart functions {#smart-functions}
```

Reviewer 2: bee43ba76

Smart functions provide a lightweight way to extend the functionality of Cumulocity across multiple components. They let you write small Javascript functions that run in a secure, isolated environment --- more powerful than configuration but much simpler than building a full microservice. For details, see [Smart function concept](/concepts/smart-function-concept/) and [Smart functions](/data-preparation/smart-functions/).

> ### Rules {#rules}

Reviewer: ```suggestion
#### Rules {#rules}
```

Reviewer 2: I think this stops them from appearing in the right menu, is that correct?

Reviewer: Yes, that is right. But structurally, they all go under "Key concepts", don´t they? (I actually missed the first one "Smart functions".). So they should go below this heading with repsect to the heading level.

Reviewer 2: Ah, yes, done: 67654248c

A rule is the deployable unit in Data Preparation. It pairs a smart function with an MQTT topic subscription and an activation state. When active, a rule processes every message posted to its subscribed topic. Rules can be created, tested with sample data, activated, deactivated, and deleted through the Data Preparation application.For details, see [Rule creation and management](/data-preparation/rule-creation-management/) and [Rule editor](/data-preparation/rule-editor/).

> ### Test data {#test-data}

Reviewer: ```suggestion
#### Test data {#test-data}
```

Reviewer 2: 67654248c
